nrf52840-hal/src/ieee802154.rs

```diff
@@ -0,0 +1,630 @@
+                        self.radio.tasks_rxen.write(|w| w.tasks_rxen().set_bit());
+                        self.wait_for_state(STATE_A::RXIDLE);
+                        return;
+                    }
+
+                    // ramping up
+                    STATE_A::RXRU => {
+                        self.wait_for_state(STATE_A::RXIDLE);
+                        return;
+                    }
+
+                    // NOTE to avoid errata 204 (see rev1 v1.4) we first go to the DISABLED state
+                    STATE_A::TXIDLE | STATE_A::TXRU => {
+                        self.radio
+                            .tasks_disable
+                            .write(|w| w.tasks_disable().set_bit());
+                        self.wait_for_state(STATE_A::DISABLED);
+                    }
+
+                    // ramping down
+                    STATE_A::RXDISABLE | STATE_A::TXDISABLE => {
+                        self.wait_for_state(STATE_A::DISABLED);
+                    }
+
+                    // cancel ongoing transfer
+                    STATE_A::RX => {
+                        self.radio.tasks_stop.write(|w| w.tasks_stop().set_bit());
+                        self.wait_for_state(STATE_A::RXIDLE);
+                        return;
+                    }
+                    STATE_A::TX => {
+                        self.radio.tasks_stop.write(|w| w.tasks_stop().set_bit());
+                        self.wait_for_state(STATE_A::TXIDLE);
+                    }
+                }
+            } else {
+                // STATE register is in an invalid state
+                unreachable!()
+            }
+        }
+    }
+
+    fn wait_for_event(&self, event: Event) {
+        match event {
+            Event::End => {
+                while self.radio.events_end.read().events_end().bit_is_clear() {
+                    continue;
+                }
+                self.radio.events_end.reset();
+            }
+            Event::PhyEnd => {
+                while self
+                    .radio
+                    .events_phyend
+                    .read()
+                    .events_phyend()
+                    .bit_is_clear()
+                {
+                    continue;
+                }
+                self.radio.events_phyend.reset();
+            }
+        }
+    }
+
+    /// Waits until the radio state matches the given `state`
+    fn wait_for_state(&self, state: STATE_A) {
+        while self.radio.state.read().state() != state {
+            continue;
+        }
+    }
+}
+
+/// NOTE must be followed by a volatile write operation
+fn dma_start_fence() {
+    atomic::compiler_fence(Ordering::Release);
+}
+
+/// NOTE must be preceded by a volatile read operation
+fn dma_end_fence() {
+    atomic::compiler_fence(Ordering::Acquire);
+}
+
+enum Event {
+    End,
+    PhyEnd,
+}
+
+/// An IEEE 802.15.4 packet
+///
+/// This `Packet` is closest to the PPDU (PHY Protocol Data Unit) defined in the IEEE spec. The API
```

Comment:
`closest to the PPDU` is ambiguous. I would state here clearly that this packet is PHR concatenated with PSDU